### PR TITLE
General fixes

### PR DESCRIPTION
**Pull request by peppy**
Optimisations and otherwise.

### CODE REVIEW

File: osu.Framework/Graphics/Drawable.cs [hunk 1]
```diff
@@ -740,12 +741,14 @@ public virtual bool Invalidate(Invalidation invalidation = Invalidation.All, Dra
                 }
 
                 alreadyInvalidated &= !drawInfoBacking.Invalidate();
+                validDrawNodes.ForEach(n => n.IsValid = false);
```
Comment (Tom94) [inline osu.Framework/Graphics/Drawable.cs]: Didn't we want to stop using .ForEach flat out?

File: osu.Framework/Graphics/Drawable.cs [hunk 2]
```diff
@@ -740,12 +741,14 @@ public virtual bool Invalidate(Invalidation invalidation = Invalidation.All, Dra
                 }
 
                 alreadyInvalidated &= !drawInfoBacking.Invalidate();
+                validDrawNodes.ForEach(n => n.IsValid = false);
                 validDrawNodes.Clear();
```
Comment (Tom94) [inline osu.Framework/Graphics/Drawable.cs]: If we are triple buffering and every Drawable has 3 corresponding DrawNodes anyway, then why don't we just make a DrawNode[3] array member inside Drawable?

File: osu.Framework/Graphics/Primitives/Triangle.cs [hunk 3]
```diff
@@ -23,23 +23,23 @@ public Triangle(Vector2 p0, Vector2 p1, Vector2 p2)
...
             if (s < 0)
-                return null;
+                return false;
 
             float t = (P0.X * P1.Y - P0.Y * P1.X + (P0.Y - P1.Y) * pos.X + (P1.X - P0.X) * pos.Y) / area2;
             if (t < 0 || (s + t) > 1)
-                return null;
+                return false;
 
-            return new Vector2(s, t);
```
Comment (Tom94) [inline osu.Framework/Graphics/Primitives/Triangle.cs]: I don't think you want to remove this. This will be needed for things like scrollbar.

Comment (peppy) [inline osu.Framework/Graphics/Primitives/Triangle.cs]: scrollbar? how so? it's already working as-is.

File: osu.Framework/Graphics/Drawable.cs (hunk 2 above)
Comment (peppy) [inline osu.Framework/Graphics/Drawable.cs]: yeah, that's the next step.

File: osu.Framework/Graphics/Drawable.cs (hunk 1 above)
Comment (peppy) [inline osu.Framework/Graphics/Drawable.cs]: habit :x

File: osu.Framework/Graphics/Primitives/Triangle.cs (hunk 3 above)
Comment (Tom94) [inline osu.Framework/Graphics/Primitives/Triangle.cs]: Sorry, I meant SliderBar. And it looks like you already had changed how it works back in osu!lazer. Nvm my comment.

**Pull request merged**